portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -117,15 +122,24 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
         // Breaks from method if the user is not logged in
         UserService userService = UserServiceFactory.getUserService();
         if (!userService.isUserLoggedIn()) {
-            response.sendRedirect(REDIRECT_URL);
+            response.sendRedirect(REDIRECT_URL_PATH);
             return;
         }
 
+        // Gets necessary data from system and request
         String userId = userService.getCurrentUser().getUserId();
         String userEmail = userService.getCurrentUser().getEmail();
-        String message = request.getParameter(DATASTORE_COMMENT_MESSAGE_PARAM);
+        String message = request.getParameter(REQUEST_MESSAGE_PARAM);
+        String languageCode = request.getParameter(REQUEST_LANGUAGE_CODE_PARAM);
         long timestamp = System.currentTimeMillis();
 
+        // Builds redirect URL based on whether the request language code is valid
+        String redirectURL = REDIRECT_URL_PATH;
+        if (languageCode != null && Arrays.asList(LANGUAGE_CODES_ARRAY).indexOf(languageCode) >= 0) {
+            redirectURL += REDIRECT_URL_QUERY + REDIRECT_URL_QUERY_LANGUAGE_PARAM + languageCode;
+        }
+        redirectURL += REDIRECT_URL_FRAGMENT;
```

Comment:
Not this time, but I would recommend to make the repeated code (e.g. make redirectURL in DataServelet and UserDataServelet) a shared method in an separate Utils class in the future.